### ISSUE TITLE
Provide ui test infrastructure for testing multi quickfixes.

### ISSUE THREAD

**Issue by miklossy** [eclipse/xtext-eclipse#1447]
It would be useful to extend the Xtext ui test infrastructure to enable the comfortable testing of multi quickfix capabilities.

**Hint:** The `org.eclipse.xtext.ui.tests.editor.quickfix.AbstractMultiQuickfixTest` already provides such basic functionality. 

- [x] Move `AbstractMultiQuickfixTest` class to the `org.eclipse.xtext.ui.testing` package.
- [x] Define test cases based on that base class for the Xtend languages. https://github.com/eclipse/xtext-xtend/issues/1061
- [x] Define test cases based on that base class for one of the Xtext examples.
- [x] Verify that the testing API is properly designed. On demand refactor the abstract base class and the implemented tests cases to eliminate duplication and to improve readability. 
- [x] Add documentation to the [UI testing section](https://www.eclipse.org/Xtext/documentation/310_eclipse_support.html#ui-testing) on the Xtext website.

**Comment by miklossy** [comment]
@drkstr101 : Thanks for offering you support on maintaining documentation, creating code samples, test cases, etc. It would be awesome if you could take care of this issue.

**Comment by drkstr101** [comment]
My pleasure! I will take a look and see what I can do.

**Comment by drkstr101** [comment]
Just a quick update...

I've freed up some time this week al will begin working on this now. I was unable to do the full build, but I believe it is just test failures, so I will move past it for now and attempt to make my changes by building the components individually. I will post another update once I've made some meaningful progress.

**Comment by drkstr101** [comment]
Alright, so I've done the first and easiest of the 3 tasks I've understood in the description.

1. This class should be moved to the org.eclipse.xtext.ui.testing package

Which leaves;

2.  The code should be cleaned up 

I will hold off on this for a bit until I have a better understanding of what needs to be cleaned up, which leaves;

3. further test cases based on that base class should be added to the Xtext/Xtend code base to verify that the API is properly designed


Since this class is responsible for testing in itself, I assume this means I should build and package a release with the relocated class, and attempt to use it to implement my own tests using that feature. This will get me over the first hump of building the whole system and bundling/installing it into a fresh workspace for testing.

Good first issue indeed!  I will report back again once I have something more useful to report.

**Comment by drkstr101** [comment]
BTW, I am sure there are developer guides laying around and I am about to take a look for them myself, but if you know of any helpful resources that are a little more difficult to find and are related more to Xtext than Eclipse in general, please do share!

**Comment by miklossy** [comment]
The [Xtext Contributing Guide](https://github.com/eclipse/xtext/blob/master/CONTRIBUTING.md#set-up-your-eclipse-workspace) describes the necessary steps on `How to set up your Xtext development environment`.

Once your development environment is properly set up, you should have both the Xtext and the Xtend repositories checked out in your workspace.

I would suggest to start implementing MultiQuickfixes test cases to the Xtend language, e.g you can add MultiQuickfix test cases for the `Triple-equals` (see the corresponding section in the `[Xtend 2.16.0 Release Notes](https://www.eclipse.org/xtend/releasenotes.html#/releasenotes/2018/12/04/version-2-16-0)`) and for the `Unnecessary Modifiers` validation rules (see the corresponding [blog post](https://blogs.itemis.com/en/xtend-2-14-unnecessary-modifiers-validation)). 

The test cases could be look like:
```
package org.eclipse.xtend.ide.tests.quickfix

import org.eclipse.xtend.ide.tests.XtendIDEInjectorProvider
import org.eclipse.xtext.testing.InjectWith
import org.eclipse.xtext.testing.XtextRunner
import org.eclipse.xtext.ui.testing.AbstractMultiQuickfixTest
import org.junit.runner.RunWith

@RunWith(XtextRunner)
@InjectWith(XtendIDEInjectorProvider)
class EqualsWithNullMultiQuickfixTest extends AbstractMultiQuickfixTest {

	// TODO: Implement test cases here
}
```
```
package org.eclipse.xtend.ide.tests.quickfix

import org.eclipse.xtend.ide.tests.XtendIDEInjectorProvider
import org.eclipse.xtext.testing.InjectWith
import org.eclipse.xtext.testing.XtextRunner
import org.eclipse.xtext.ui.testing.AbstractMultiQuickfixTest
import org.junit.runner.RunWith

@RunWith(XtextRunner)
@InjectWith(XtendIDEInjectorProvider)
class RemoveUnnecessaryModifiersMultiQuickfixTest extends AbstractMultiQuickfixTest {
	
	// TODO: Implement test cases here
	
}
```

> 
I assume this means I should build and package a release with the relocated class, and attempt to use it to implement my own tests using that feature. 

You do not need to build and package a release with the relocated class in order to use it in the Xtend repository. Once this issue is fixed and get merged to the master branch, you can bind the Xtext latest version into your target platform and so you can test it with further Xtext-based projects.

**Comment by drkstr101** [comment]
Thank you so much for taking the time to give me such detailed and helpful info! I will start digesting it all so I can start working on the next step right away.

Cheers!

**Comment by drkstr101** [comment]
@miklossy I was definitely doing things the hard way before! Now I am cooking with gas, using the very handy Contributes bundle to make my changes. I was doing everything manually before. Oops!

I should be able to get a couple test cases in there for you soon.

Cheers!

**Comment by drkstr101** [comment]
@miklossy The next step will take place in `xtext-xtend`, so I have submitted a PR for the first step to `xtext-eclipse` as https://github.com/eclipse/xtext-eclipse/pull/1464

Now I will use this to add `EqualsWithNullMultiQuickfixTest` and `RemoveUnnecessaryModifiersMultiQuickfixTest` to `xtext-xtend`. More to follow.

Cheers!

**Comment by kthoms** [comment]
@drkstr101 @miklossy Can this considered fixed with the merged PR?

**Comment by miklossy** [comment]
@kthoms The merged PR was only the first step towards the entire fix. I updated the description to better reflect all the things that needs to be done before this issue can be considered as fixed.

**Comment by miklossy** [comment]
I have just added the documentation about the `AbstractMultiQuickfixTest` base class to the [UI testing section](https://www.eclipse.org/Xtext/documentation/310_eclipse_support.html#ui-testing) on the Xtext website.

**Comment by miklossy** [comment]
@drkstr101 : Thank you very much for your contribution!

Fixed in 2.23
